crates/sound/src/audio_backends/pw_backend.rs

```diff
@@ -4,25 +4,179 @@
+}
+
+pub struct PWVoice {
+    pub ring: spa_ringbuffer,
+    pub buffer: [u8; RINGBUFFER_SIZE as usize],
+    // TODO: stream shall be protected with Arc<Mutex<T>> but this would
+    // require that pipewire::stream::Stream is Send which is not the
+    // the case currently. For the moment, use Mutex and impl Send and Sync
+    pub stream: Stream<i32>,
+}
+
+// SAFETY: Safe as the structure can be sent to another thread.
+unsafe impl Send for PwBackend {}
+
+// SAFETY: Safe as the structure can be shared with another thread as the state
+// is protected with a lock.
+unsafe impl Sync for PwBackend {}
+
+pub struct PwBackend {
+    pub streams: PWVoice
+}
 
 impl PwBackend {
     pub fn new() -> Self {
-        PwBackend {}
+        pw::init();
+
+        let buffer = [0; RINGBUFFER_SIZE as usize];
+
+        let ring= spa_ringbuffer {
+            readindex : 0 as u32,
+            writeindex : 0 as u32,
+        };
+
+        let wrap_mainloop = WrapMainLoop {
+            mainloop : pw::MainLoop::new().expect("we can't create mainloop")
+        };
+
+        let stream = pw::stream::Stream::<i32>::with_user_data(
+            &wrap_mainloop.mainloop,
+            "audio-output",
+            properties! {
+                *pw::keys::MEDIA_TYPE => "Audio",
+                *pw::keys::MEDIA_CATEGORY => "Playback",
+                *pw::keys::MEDIA_ROLE => "Music",
+            },
+            0,
+        )
+        .process(|stream, _user_data| {
+            // TODO: to work this
+            unsafe {
+                let raw_buffer = stream.dequeue_raw_buffer();
+                let datas = raw_buffer.as_mut();
+
+                // TODO: to put some data
+
+                stream.queue_raw_buffer(raw_buffer);
+            }
+        })
+        .state_changed(|old, new| {
+            println!("State changed: {:?} -> {:?}", old, new);
+        })
+        .create().expect("we can't create the stream");
+
+
+        // TODO: to set this during PCM_SET
+        // this does not look safe
+        // see https://users.rust-lang.org/t/cast-c-int-to-mut-c-void/44766
+        let mut buff = [0; 1024];
+        let p_buff = &mut buff as *mut i32 as *mut c_void;
+
+        let mut b = spa_pod_builder {
+            data : p_buff,
+            size: buff.len() as u32,
+            _padding : 0,
+            callbacks : spa_callbacks {
+                funcs: std::ptr::null(),
+                data: std::ptr::null_mut()
+            },
+            state: spa_pod_builder_state {
+                offset : 0,
+                flags: 0,
+                frame: std::ptr::null_mut()
+            }
+        };
+
+        let mut pos: [u32; 64] = [0; 64];
+
+        pos[0] = SPA_AUDIO_CHANNEL_FL;
+        pos[1] = SPA_AUDIO_CHANNEL_FR;
+
+        let mut info = spa_audio_info_raw {
+            format : SPA_AUDIO_FORMAT_S16,
+            rate : 44100,
+            flags : 0,
+            channels : 2,
+            position : pos
+        };
```

Comment:
I somehow feel like these types could possibly get a proper wrapper and possibly upstreamed later

Reply:
this is just a hardcoded placeholder for now, we should get this data from the VIRTIO_SND_R_PCM_SET_PARAMS request type to be used in the method that calls the prepare phase.

Reply:
I meant the construction of the ffi structure here, not the passed data.

Reply:
Do you mean to have a some sort of macro to instantiate it and set the initial values?

Reply:
Cool, thanks.